Docs/StyleGuide/Style-Guide.md

```diff
@@ -0,0 +1,280 @@
+definitions. Each method specification clause should be in a separate line.
+
+In case the Method signature is too long, we can break it down.
+```dafny
+method {:<attributes>} MethodName(param1: Type, param2: Type,
+        param3: Type, param4: Type, param5: Type)
+    returns (ret1: Type, ret2: Type, ret3: Type, ret4: Type,
+        ret5: Type)
+    decreases param1
+    requires P1()
+    requires P2()
+    requires P3()
+    modifies param2
+    modifies param3
+    ensures Q1()
+    ensures Q2()
+```
+
+Multiple `requires` or `ensures` can be combined into one:
+```dafny
+requires
+    P1() &&
+    P2() &&
+    P3()
+```
+The same rules apply to `function`, `predicate`, and `lemma` definitions.
+
+## Things to Avoid
+
+### Parenthesis
+
+In a lot of cases, Dafny does not require parentheses around expressions. Here are some examples.
+
+* If-Else-While Statements
+```dafny
+// YES
+var i := 1;
+while i < 10 {
+    ...
+    if 1 < i {
+        ...
+    }
+    ...
+}
+
+// NO
+var i := 1;
+while (i < 10) {
+    ...
+    if (1 < i) {
+        ...
+    }
+    ...
+}
+```
+
+* Simple Boolean/Arithmetic Expressions
+```dafny
+// YES
+method Collatz(num: nat)
+    decreases *
+{
+    var n := num;
+    while 1 < n
+        decreases *
+    { 
+        n := if n % 2 == 0 then n / 2 else n * 3 + 1;
+    }
+}
+
+// NO
+method Collatz(num: nat)
+    decreases *
+{
+    var n := num;
+    while (1 < n) // unnecessary parentheses
+        decreases *
+    { 
+        n := if ((n % 2) == 0) then (n / 2) else ((n * 3) + 1); // unnecessary parentheses
+    }
+}
+```
+### Whitespaces
+
+Avoid unnecessary whitespaces inside expressions.
+
+#### Type Declaration
+Type declaration should have a form of `variableName: variableType`.
+```dafny
+// YES
+var i: int := 1;
+class {:extern} Util {
+    var {:extern} Exception: System.String
+}
+
+// NO
+var i : int := 1; // unnecessary whitespaces
+class {:extern} Util {
+    var {:extern} Exception : System.String // unnecessary whitespaces
+}
+```
+
+#### Function, Method, Predicate, and Lemma Declaration
+The `function`, `method`, `predicate`, and `lemma` definitions should have a form of
+`FunctionName(parameterName: parameterType, ...)`.
+
+```dafny
+// YES
+function method Foo<int>(i: int): int
+
+// NO
+function method Foo<int> (i : int) : int // unnecessary whitespaces
+```
+
+Avoid too little or too many whitespaces that reduce the overall readability.
+
+```dafny
+// YES
+lemma MyLemma<A, B>(x: seq<seq<A>>, y: B) {
+    ...
+}
+
+// NO
+lemma MyLemma <A,B> ( x : seq<seq<A>> , y :B){
+    ...
+}
+```
+
+## Recommendations
+This section describes a few recommendations that can help make code more readable and easy to follow, although not
+strictly enforced.
+
+### Externs
+Try to name them the same in Dafny and C# whenever possible, so that in Dafny we only have to write `{:extern}`, not
+`{:extern "<name>"}`.
```

Comment:
Too strict. Dafny doesn't target only C#.

Reply:
Changed it to "target language".